src/cgutils.cpp

```diff
@@ -1148,6 +1198,8 @@ static jl_cgval_t typed_load(Value *ptr, Value *idx_0based, jl_value_t *jltype,
     //else {
         Instruction *load = builder.CreateAlignedLoad(data, isboxed ?
             alignment : julia_alignment(data, jltype, alignment), false);
+        if (isboxed)
+            load = maybe_mark_load_dereferenceable(load, true, jltype);
```

Comment:
The deferenceability attribute applies to the loaded value (i.e. the loaded pointer is dereferencable, not the pointer we're loading to) and includes the size of the object. It would not allow hoisting the load past a bounds check, unless the pointer we're loading from itself was annotate as dereferenceable. However, since we annotate it with the size, even then it would only be able to do that if the pointer actually were in bounds.